crmsh/bootstrap.py

```diff
@@ -1202,6 +1202,84 @@ def generate_pacemaker_remote_auth():
     utils.chmod(PCMK_REMOTE_AUTH, 0o640)
 
 
+class FirewallManager:
+
+    SERVICE_NAME = "high-availability"
+
+    def __init__(self, peer=None):
+        self.shell = None
+        self.peer = peer
+        self.firewalld_running = False
+        self.firewall_cmd = None
+        self.firewall_cmd_prompt_option = ""
+        self.peer_msg = ""
+        self.firewalld_installed = utils.package_is_installed("firewalld")
+
+        if self.firewalld_installed:
+            self.shell = sh.cluster_shell()
+            rc, _, _ = self.shell.get_rc_stdout_stderr_without_input(self.peer, "firewall-cmd --state")
+            self.firewalld_running = rc == 0
+            self.firewall_cmd = "firewall-cmd" if self.firewalld_running else "firewall-offline-cmd"
+            self.firewall_cmd_prompt_option = " --permanent" if self.firewalld_running else ""
+            self.peer_msg = f"on {self.peer}" if self.peer else f"on {utils.this_node()}"
+
+    def _service_is_available(self) -> bool:
+        cmd = f"{self.firewall_cmd} --info-service={self.SERVICE_NAME}"
+        rc, _, _ = self.shell.get_rc_stdout_stderr_without_input(self.peer, cmd)
+        if rc != 0:
+            logger.warning("Firewalld service %s is not available %s", self.SERVICE_NAME, self.peer_msg)
+            return False
+        return True
+
+    def _reload_firewalld(self):
+        if not self.firewalld_running:
+            return
+        self.shell.get_rc_stdout_stderr_without_input(self.peer, "firewall-cmd --reload")
+
+    def add_service(self):
+        if not self.firewalld_installed or not self._service_is_available():
+            return
+        cmd = f"{self.firewall_cmd}{self.firewall_cmd_prompt_option} --add-service={self.SERVICE_NAME}"
+        rc, _, err = self.shell.get_rc_stdout_stderr_without_input(self.peer, cmd)
+        if rc != 0:
+            logger.error("Failed to add firewalld service %s %s: %s", self.SERVICE_NAME, self.peer_msg, err)
+            return
+        logger.info("Added firewalld service %s %s", self.SERVICE_NAME, self.peer_msg)
+        self._reload_firewalld()
```

Comment:
Should not reload firewalld, as this may cause potential problems if users have make some runtime configuration.

In firewall-cmd(1):
> all runtime only changes done until reload are lost with reload if they have not been also in permanent configuration.

> If you want to make a change in runtime and permanent configuration, use the same call with and without the --permanent option.

Reply:
Thanks for reminding! So, use `--permanent` should be enough.